docker/docker_image_src.go

```diff
@@ -30,15 +32,68 @@ type dockerImageSource struct {
 
 // newImageSource creates a new ImageSource for the specified image reference.
 // The caller must call .Close() on the returned ImageSource.
-func newImageSource(sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
-	c, err := newDockerClientFromRef(sys, ref, false, "pull")
+func newImageSource(ctx context.Context, sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
+	registry, err := sysregistriesv2.FindRegistry(sys, ref.ref.Name())
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+
+	if registry == nil {
+		// No registry was found be the provided reference, so we create a
+		// fallback registry by hand to make the client creation below work as
+		// intended.
+		registry = &sysregistriesv2.Registry{
+			Endpoint: sysregistriesv2.Endpoint{
+				URL: ref.ref.String(),
+			},
+		}
+	}
+
+	// Found the registry within the sysregistriesv2 configuration. Now we test
+	// all endpoints for the manifest availability. If an working image source
+	// was found, it will be used for all future pull actions.
+	var (
+		imageSource     *dockerImageSource
+		manifestLoadErr error
+	)
+	for _, endpoint := range append(registry.Mirrors, registry.Endpoint) {
+		newRef, err := rewriteReference(ref, registry.Prefix, endpoint.URL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error rewriting reference")
+		}
+
+		client, err := newDockerClientFromRef(sys, newRef, false, "pull")
+		client.endpointSkipVerify = endpoint.Insecure
+		if err != nil {
+			return nil, err
+		}
+
+		testImageSource := &dockerImageSource{
+			ref: newRef,
+			c:   client,
+		}
+
+		manifestLoadErr = testImageSource.ensureManifestIsLoaded(ctx)
+		if manifestLoadErr == nil {
+			imageSource = testImageSource
+			break
+		}
 	}
-	return &dockerImageSource{
-		ref: ref,
-		c:   c,
-	}, nil
+
+	return imageSource, manifestLoadErr
+}
+
+// rewriteReference will substitute the provided reference prefix `from` to the
+// given `to` from the `ref` and creates a new reference from it. The function
+// errors if the newly created reference is not parsable.
+func rewriteReference(ref dockerReference, from, to string) (dockerReference, error) {
```

Comment:
I think this should be a method of `registry.Endpoint` which should also make the comment easier to brain parse.

The reasoning to move it to the `sysregistriesv2` package is to keep it self-contained. Having it here means that a core part of the package's functionality is scattered making it harder to maintain and read, etc.

Reply:
I am also not sure if the current implementation is correct. Comparing it to the one of the Moby PR shows quite some differences (e.g., special casing the docker registry):

https://github.com/moby/moby/pull/34319/files#diff-73caf4b121449e3cc9df8ab7554b31baR70

Reply:
> I am also not sure if the current implementation is correct. Comparing it to the one of the Moby PR shows quite some differences (e.g., special casing the docker registry):
> 
> https://github.com/moby/moby/pull/34319/files#diff-73caf4b121449e3cc9df8ab7554b31baR70

This should be already handled by `reference.ParseNamed`. That's also why I did not add certain tests for it. Does that make sense?